src/client/serialize/daos_serialize.c

```diff
@@ -1175,3 +1245,1833 @@ daos_cont_deserialize_attrs(char *filename, uint64_t *num_attrs,
+				single_val = &(*hdf5->ak)[*akey_index].rec_single_val;
+				single_val->p = NULL;
+				single_val->len = 0;
+
+				/* create a record dset only for array types */
+				memset(&rec_name, 0, rec_name_len);
+				path_len = snprintf(rec_name, rec_name_len, "%d", *akey_index);
+				if (path_len > rec_name_len) {
+					D_ERROR("Record name too long "DF_RC"\n", DP_RC(rc));
+					D_GOTO(out, rc);
+				}
+				rc = serialize_recx_array(hdf5, &diov, &aiov, rec_name, akey_index,
+							  oh, &iod, bytes_read);
+				if (rc != 0) {
+					D_ERROR("Failed to serialize recx array "DF_RC"\n",
+						DP_RC(rc));
+					D_GOTO(out, rc);
+				}
+			} else {
+				rc = serialize_recx_single(hdf5, &diov, oh, &iod, akey_index,
+							   bytes_read);
+				if (rc != 0) {
+					D_ERROR("Failed to serialize recx single "DF_RC"\n",
+						DP_RC(rc));
+					D_GOTO(out, rc);
+				}
+			}
+			/* advance to next akey returned */
+			akey_ptr += akey_kds[j].kd_key_len;
+			(*akey_index)++;
+			D_FREE(akey);
+		}
+		*total_akeys += akey_number;
+	}
+out:
+	if (rc != 0)
+		D_FREE(akey);
+	D_FREE(small_key);
+	D_FREE(large_key);
+	return rc;
+}
+
+static int
+serialize_kv_rec(struct hdf5_args *hdf5, daos_key_t dkey, daos_handle_t *oh, int *dkey_index,
+		 char *dkey_val, uint64_t *bytes_read)
+{
+	void		*buf = NULL;
+	int		rc;
+	hvl_t		*kv_val;
+	daos_size_t	size = 0;
+
+	/* get the size of the value */
+	rc = daos_kv_get(*oh, DAOS_TX_NONE, 0, dkey_val, &size, buf, NULL);
+	if (rc != 0) {
+		D_ERROR("Failed to fetch KV object "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+	D_ALLOC(buf, size);
+	if (buf == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+
+	rc = daos_kv_get(*oh, DAOS_TX_NONE, 0, dkey_val, &size, buf, NULL);
+	if (rc != 0) {
+		D_ERROR("Failed to fetch KV object "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out, rc);
+	}
+
+	(*bytes_read) += size;
+	kv_val = &(*hdf5->dk)[*dkey_index].rec_kv_val;
+	D_ALLOC(kv_val->p, (uint64_t)size);
+	if (kv_val->p == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	memcpy(kv_val->p, buf, (uint64_t)size);
+	kv_val->len = (uint64_t)size;
+out:
+	D_FREE(buf);
+	return rc;
+}
+
+static int
+serialize_dkeys(struct hdf5_args *hdf5, daos_obj_id_t oid, daos_handle_t coh, int *oid_index,
+		int *dkey_index, int *akey_index, daos_handle_t *oh, bool is_kv,
+		int *total_dkeys, int *total_akeys, uint64_t *bytes_read)
+{
+	int		rc = 0;
+	int		i = 0;
+	daos_anchor_t	dkey_anchor = {0};
+	d_sg_list_t	dkey_sgl;
+	d_iov_t		dkey_iov;
+	daos_key_desc_t	dkey_kds[ENUM_DESC_NR] = {0};
+	uint32_t	dkey_number = ENUM_DESC_NR;
+	char		*dkey = NULL;
+	char		*dkey_ptr;
+	daos_key_t	diov;
+	int		path_len = 0;
+	hvl_t		*dkey_val;
+	char		key_val[ENUM_KEY_BUF];
+	char		*small_key = NULL;
+	char		*large_key = NULL;
+	char		*key_buf = NULL;
+	daos_size_t	key_buf_len = 0;
+
+	D_ALLOC(small_key, ENUM_DESC_BUF);
+	if (small_key == NULL)
+		D_GOTO(out, rc = -DER_NOMEM);
+	D_ALLOC(large_key, ENUM_LARGE_KEY_BUF);
+	if (large_key == NULL)
```

Comment:
@daltonbohning this was the way the other test code does it when I looked, but probably to use less memory since the large buffer isn't always necessary. Not sure which is better choice or how often a "large" buffer is needed.

Reply:
But if `large_key` is always allocated upfront, then it doesn't actually use less memory, right? Since it's allocated once here, and then the actual key is just copied, no realloc'ed

Reply:
@daltonbohning yeah true. I should have had it only allocating if the key was too big. I moved that allocation.